### PR TITLE
python fixes

### PR DESCRIPTION
**Pull request by NickeZ**
This is a pretty big reorganization of the python stuff. Tested everything with bb02 and bb02btc firmware.

* Reorganize
* Merge similar code into one code base
* Package bitbox02 module
* Add support for btc-only
* Add support for bootloader (boot command)

### CODE REVIEW

File: py/load_firmware.py [hunk 1]
```diff
@@ -48,43 +66,58 @@ def main() -> int:
...
-        bitbox = BitBox02(device_info=bitboxes[0], show_pairing_callback=show_pairing)
+        bitbox = BitBox02(device_info=bitbox, show_pairing_callback=_show_pairing)
         bitbox.reboot()
-        bootloaders = _wait_for_bootloaders()
 
-    if len(bootloaders) > 1:
-        eprint("Multiple bootloaders detected. Only one supported")
-        return 1
+        # wait for it to reboot
+        while True:
```
Comment (benma) [inline py/load_firmware.py]: if you have multiple it might take the wrong one

File: py/load_firmware.py [hunk 2]
```diff
@@ -48,43 +66,58 @@ def main() -> int:
...
+
+    if bootloader is None:
+        try:
+            bootloader = devices.get_bitbox02btc_bootloader()
+        except TooManyFoundException:
+            eprint("Found multiple bb02 bootloader btc editions. Only one supported.")
+
+    if bootloader is None:
+        bitbox = None
+        try:
```
Comment (benma) [inline py/load_firmware.py]: a bit strange semantics, it will work if one device is plugged in, but also if 2 + 1 are plugged in.

What about having a function which returns all devices of all editions? The distinction is usually not important, and if it is we can have a `@property edition` to query for it per device/bootloader.

**Comment by benma** [comment]
It would be much easier to review if this many commits with a description per commit.

With big PRs I tend to not look very carefully as it's a bit overwhelming. Manual testing is much harder because I don't know where to focus.

Here even multiple PRs would have helped, as some seems unrelated (like the packaging).

**Comment by NickeZ** [comment]
> It would be much easier to review if this many commits with a description per commit.
> 
> With big PRs I tend to not look very carefully as it's a bit overwhelming. Manual testing is much harder because I don't know where to focus.
> 
> Here even multiple PRs would have helped, as some seems unrelated (like the packaging).

The goal was to do the packaging but when I started to shave that jak I ended up modifying much more than expected. At least I split out the packaging specific stuff in a separate commit.

File: py/load_firmware.py (hunk 2 above)
Comment (NickeZ) [reply py/load_firmware.py]: ack

File: py/load_firmware.py (hunk 1 above)
Comment (NickeZ) [reply py/load_firmware.py]: ack

**Comment by NickeZ** [comment]
There is still some work to be done here. mypy linter currently fails.

**Comment by NickeZ** [comment]
Fixed the linter issues on the flight yesterday and I tested send_message and load_firmware. I think this is ready to be merged.

File: scripts/lint-python [hunk 3]
```diff
@@ -28,7 +28,7 @@ fi
 # behavior).
 # We must typecheck the whole `py` directory even if only a few files are modified so that mypy
 # sees all types
-${MYPY} --implicit-reexport --namespace-packages --ignore-missing-imports --strict py
+${MYPY} --implicit-reexport --namespace-packages --ignore-missing-imports --strict py py/bitbox02
```
Comment (benma) [inline scripts/lint-python]: why is this needed? doesn't it do `py` and all imported dependencies?

File: py/bitbox02/bitbox02/bitbox02.py [hunk 4]
```diff
@@ -269,6 +272,9 @@ def _msg_query(
             print(response)
         return response
 
+    def u2f_query(self, msg: bytes) -> bytes:
```
Comment (benma) [inline py/bitbox02/bitbox02/bitbox02.py]: What's the purpose of this? Seems like it ca be deleted as it's u2f, and empty body.

File: py/bitbox02/bitbox02/devices.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    """
+    devices = get_any_bitboxes()
+    if len(devices) > 1:
+        raise TooManyFoundException(len(devices))
+    if not devices:
+        raise NoneFoundException()
+    return devices[0]
+
+
+def get_any_bitbox_bootloaders() -> List[DeviceInfo]:
```
Comment (benma) [inline py/bitbox02/bitbox02/devices.py]: `bitbox02`

File: py/bitbox02/bitbox02/devices.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    Searches for both btc-only and non-btc-only devices
+    Returns:
+        List of devices
+    """
+    devices = get_bitbox02_bootloaders()
+    devices.extend(get_bitbox02btc_bootloaders())
+    return devices
+
+
+def get_any_bitbox_bootloader() -> DeviceInfo:
```
Comment (benma) [inline py/bitbox02/bitbox02/devices.py]: same, `bitbox02`

File: py/bitbox02/setup.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    long_description_content_type="text/markdown",
+    url="https://github.com/digitalbitbox/bitbox02-firmware",
+    python_requires=">=3.6",
+    classifiers=[
+        "Intended Audience :: Developers",
+        "License :: OSI Approved :: Apache Software License",
+        "Programming Language :: Python :: 3.6",
+    ],
+    keywords="digitalbitbox bitbox bitbox02 bitcoin litecoin ethereum erc20 u2f",
+    install_requires=["hidapi", "noiseprotocol", "protobuf", "ecdsa", "semver"],
```
Comment (benma) [inline py/bitbox02/setup.py]: this is missing version bounds like in requirements.py. I think the newest protobuf does not work yet with our code, for example.

Speaking of, is `requirements.py` still needed?

File: scripts/lint-python (hunk 3 above)
Comment (NickeZ) [reply scripts/lint-python]: It's needed because py/bitbox02 is a package now. So the imports in for example `send_message.py` requires you to actually install bitbox02 before you can use it.

**Comment by NickeZ** [comment]
> I think the linter is not working properly. I can't run `./py/send_message.py`:
> 
> ```
> Traceback (most recent call last):
>   File "./py/send_message.py", line 25, in <module>
>     from bitbox02 import HARDENED
> ImportError: cannot import name 'HARDENED'
> ```
> 
> Running `./scripts/lint-python py/send_message.py` gives a lot of similar issues.

You need to install bitbox02. Run `python3 -m pip install -e .` in the `py/bitbox02` directory.

File: py/bitbox02/setup.py (hunk 7 above)
Comment (NickeZ) [reply py/bitbox02/setup.py]: Oh, I didn't know that you could put versions in the `setup.py` nowadays.

After reading `https://packaging.python.org/discussions/install-requires-vs-requirements/` it seems like we can remove the requirements files yes.

File: py/bitbox02/bitbox02/bitbox02.py (hunk 4 above)
Comment (NickeZ) [reply py/bitbox02/bitbox02/bitbox02.py]: ack

File: py/bitbox02/bitbox02/devices.py (hunk 5 above)
Comment (NickeZ) [reply py/bitbox02/bitbox02/devices.py]: ack

**Comment by NickeZ** [comment]
Force pushed a new version
